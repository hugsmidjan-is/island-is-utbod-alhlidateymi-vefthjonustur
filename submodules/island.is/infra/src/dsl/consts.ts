import { resolve } from 'path'

export const rootDir = resolve(__dirname, '..', '..', '..')

export const COMMON_SECRETS = {
  CONFIGCAT_SDK_KEY: '/k8s/configcat/CONFIGCAT_SDK_KEY',
}
