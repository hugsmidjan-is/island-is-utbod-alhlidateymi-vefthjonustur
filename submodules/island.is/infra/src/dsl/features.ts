export { ServerSideFeature as FeatureNames } from '../../../libs/feature-flags/src/lib/features'
