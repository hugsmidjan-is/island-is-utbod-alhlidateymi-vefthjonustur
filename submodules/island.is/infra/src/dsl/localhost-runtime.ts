import { ServiceDefinition, ServiceDefinitionCore } from './types/input-types'
import { ReferenceResolver } from './types/charts'
import { getMockName, hostPortNumber } from './mocks/mocks-support'

export class Localhost implements ReferenceResolver {
  deps: { [name: string]: Set<string> } = {}
  ports: { [name: string]: number } = {}
  mocks: { [name: string]: string } = {}

  ref(from: ServiceDefinitionCore, to: ServiceDefinition | string) {
    if (typeof to === 'object') {
      this.ports[to.name] = this.ports[to.name] ?? hostPortNumber(to.name)
      const dependencies = this.deps[to.name] ?? new Set<string>()
      this.deps[to.name] = dependencies.add(from.name)
      return `localhost:${this.ports[to.name]}`
    } else {
      const { name, host } = getMockName(to)
      this.ports[name] = this.ports[name] ?? hostPortNumber(host)
      this.mocks[name] = host
      return `http://localhost:${this.ports[name]}`
    }
  }
}
