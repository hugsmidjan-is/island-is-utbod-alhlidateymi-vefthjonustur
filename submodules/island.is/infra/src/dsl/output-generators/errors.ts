export const checksAndValidations = (serviceName: string) => {
  let allErrors: string[] = []
  const checkCollisions = (
    target: { [name: string]: string },
    source: { [name: string]: string },
  ) => {
    const targetKeys = Object.keys(target)
    addToErrors(
      Object.keys(source)
        .filter((srcKey) => targetKeys.includes(srcKey))
        .map(
          (key) =>
            `Collisions in ${serviceName} for environment or secrets for key ${key}`,
        ),
    )
  }
  const mergeObjects = (
    target: { [name: string]: string },
    source: { [name: string]: string },
  ) => {
    checkCollisions(target, source)
    Object.assign(target, source)
  }
  const addToErrors = (errors: string[]) => {
    allErrors.push(...errors)
  }
  const getErrors = () => {
    return [...allErrors]
  }

  return {
    addToErrors,
    mergeObjects,
    checkCollisions,
    getErrors,
  }
}
