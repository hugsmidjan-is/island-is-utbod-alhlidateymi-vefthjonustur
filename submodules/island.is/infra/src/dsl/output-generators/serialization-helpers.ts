import {
  EnvironmentVariablesForEnv,
  ServiceDefinitionForEnv,
  ValueSource,
} from '../types/input-types'
import { ReferenceResolver, EnvironmentConfig } from '../types/charts'
import { ContainerEnvironmentVariables } from '../types/output-types'

export const resolveWithMaxLength = (str: string, max: number) => {
  if (str.length > max) {
    return `${str.substr(0, Math.ceil(max / 3))}${str.substr((-max / 3) * 2)}`
  }
  return str
}

export function serializeValueSource(
  value: ValueSource,
  deployment: ReferenceResolver,
  service: ServiceDefinitionForEnv,
  env: EnvironmentConfig,
): { type: 'success'; value: string } {
  const result =
    typeof value === 'string'
      ? value
      : value({
          env: env,
          featureDeploymentName: env.feature,
          svc: (dep) =>
            deployment.ref(
              service,
              typeof dep === 'string' ? dep : dep.serviceDef,
            ),
        })
  return { type: 'success', value: result }
}

export function serializeEnvironmentVariables(
  service: ServiceDefinitionForEnv,
  deployment: ReferenceResolver,
  envs: EnvironmentVariablesForEnv,
  env: EnvironmentConfig,
): { envs: ContainerEnvironmentVariables } {
  return Object.entries(envs).reduce(
    (acc, [name, value]) => {
      const r = serializeValueSource(value, deployment, service, env)
      switch (r.type) {
        case 'success':
          return {
            envs: {
              ...acc.envs,
              [name]: r.value,
            },
          }
      }
    },
    { envs: {} as ContainerEnvironmentVariables },
  )
}

export const postgresIdentifier = (id: string) => id.replace(/[\W\s]/gi, '_')
