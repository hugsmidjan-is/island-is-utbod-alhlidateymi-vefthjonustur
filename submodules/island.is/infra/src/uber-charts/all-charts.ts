import { Services as ISServices } from './islandis'
import { Services as JSServices } from './judicial-system'
import { Services as IDSServices } from './identity-server'
import { EnvironmentServices, OpsEnvName } from '../dsl/types/charts'
import { OpsEnv } from '../dsl/types/input-types'

export type ChartName = 'islandis' | 'judicial-system' | 'identity-server'
export const ChartNames: ChartName[] = [
  'islandis',
  'judicial-system',
  'identity-server',
]
export const OpsEnvNames: OpsEnv[] = ['dev', 'staging', 'prod']
export const Charts: { [name in ChartName]: EnvironmentServices } = {
  'identity-server': IDSServices,
  islandis: ISServices,
  'judicial-system': JSServices,
}

export const Deployments: {
  [name in ChartName]: { [env in OpsEnv]: OpsEnvName }
} = {
  'judicial-system': { dev: 'dev01', staging: 'staging01', prod: 'prod' },
  islandis: { dev: 'dev01', staging: 'staging01', prod: 'prod' },
  'identity-server': { dev: 'devIds', staging: 'stagingIds', prod: 'prod-ids' },
}
