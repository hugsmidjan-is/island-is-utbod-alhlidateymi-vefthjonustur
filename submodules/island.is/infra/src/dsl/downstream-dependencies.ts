import { ServiceDefinition, ServiceDefinitionCore } from './types/input-types'
import { ReferenceResolver, EnvironmentConfig } from './types/charts'
import cloneDeep from 'lodash/cloneDeep'
import { generateOutput } from './processing/rendering-pipeline'
import { renderers } from './upstream-dependencies'
import { ServiceBuilder } from './dsl'

const MAX_LEVEL_DEPENDENCIES = 20

class DownstreamDependencyTracer implements ReferenceResolver {
  deps: { [name: string]: Set<string> } = {}

  ref(from: ServiceDefinitionCore, to: ServiceDefinition | string) {
    if (typeof to === 'object') {
      const dependecies = this.deps[to.name] ?? new Set<string>()
      this.deps[to.name] = dependecies.add(from.name)
      return 'tracer'
    } else {
      return to
    }
  }
}

const findDownstreamDependencies = (
  runtime: DownstreamDependencyTracer,
  svc: ServiceDefinition | string,
  level: number = 0,
): string[] => {
  const deps = runtime.deps[typeof svc === 'string' ? svc : svc.name]
  if (level > MAX_LEVEL_DEPENDENCIES)
    throw new Error(
      `Too deep level of dependencies - ${MAX_LEVEL_DEPENDENCIES}. Some kind of circular dependency or you fellas have gone off the deep end ;)`,
    )
  const currentDeps = Array.from(deps ?? new Set<string>())
  if (currentDeps) {
    const serviceDependencies = currentDeps
    return serviceDependencies
      .map((dependency) =>
        findDownstreamDependencies(runtime, dependency, level + 1),
      )
      .flatMap((x) => x)
      .concat(serviceDependencies)
  } else {
    return []
  }
}

export const getWithDownstreamServices = async (
  env: EnvironmentConfig,
  habitat: ServiceBuilder<any>[],
  services: ServiceBuilder<any>[],
): Promise<ServiceBuilder<any>[]> => {
  const dependencyTracer = new DownstreamDependencyTracer()
  const localHabitat = cloneDeep(habitat)
  await generateOutput({
    runtime: dependencyTracer,
    services: localHabitat,
    outputFormat: renderers.helm,
    env: env,
  }) // doing this so we find out the dependencies
  const downstreamServices = services
    .map((s) => findDownstreamDependencies(dependencyTracer, s.serviceDef))
    .flatMap((x) => x)
  return services
    .concat(
      downstreamServices.map(
        (dep) => habitat.find((s) => s.serviceDef.name === dep)!,
      ),
    )
    .reduce(
      (
        acc: ServiceBuilder<any>[],
        cur: ServiceBuilder<any>,
      ): ServiceBuilder<any>[] =>
        acc.indexOf(cur) === -1 ? acc.concat([cur]) : acc,
      [],
    )
}
