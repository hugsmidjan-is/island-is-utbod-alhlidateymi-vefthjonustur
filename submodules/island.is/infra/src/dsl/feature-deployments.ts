import { getWithDownstreamServices } from './downstream-dependencies'
import { ServiceBuilder } from './dsl'
import { EnvironmentConfig } from './types/charts'

export const getFeatureAffectedServices = async (
  habitat: ServiceBuilder<any>[],
  services: ServiceBuilder<any>[],
  excludedServices: ServiceBuilder<any>[] = [],
  env: EnvironmentConfig,
) => {
  const feature = env.feature
  if (typeof feature !== 'undefined') {
    const excludedServiceNames = excludedServices.map((f) => f.serviceDef.name)

    const affectedServices = (
      await getWithDownstreamServices(env, habitat, services)
    ).reduce(
      (acc, f) =>
        excludedServiceNames.includes(f.serviceDef.name)
          ? { included: acc.included, excluded: [...acc.excluded, f.name()] }
          : { included: [...acc.included, f], excluded: acc.excluded },
      {
        included: [] as ServiceBuilder<any>[],
        excluded: [] as string[],
      },
    )

    return affectedServices
  } else {
    throw new Error('Feature deployment with a feature name not defined')
  }
}
