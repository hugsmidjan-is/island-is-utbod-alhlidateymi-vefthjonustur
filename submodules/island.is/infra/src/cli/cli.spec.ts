import { renderSecretsCommand } from './render-secrets'
import { Charts } from '../uber-charts/all-charts'

import { service, ServiceBuilder } from '../dsl/dsl'
import { SecretOptions } from '../dsl/output-generators/map-to-localrun'
import { renderLocalServices } from './render-local-mocks'

jest.mock('../uber-charts/all-charts')

jest.mock('../dsl/consts', () => ({
  rootDir: '.',
  COMMON_SECRETS: {},
}))

jest.spyOn(console, 'log').mockImplementation(() => {})
jest.spyOn(console, 'error')

jest.mock('@aws-sdk/client-ssm', () => ({
  SSM: jest.fn(() => ({
    getParameters: jest.fn(() => ({
      Parameters: [
        {
          Name: '/k8s/my-service/A',
          Value: 'B',
        },
        {
          Name: '/k8s/my-service/REMOTE_URL',
          Value: 'https://www.website.tld/some/slug-path',
        },
        {
          Name: '/k8s/my-service/JSON_SECRET',
          Value: '{"key": ["value"]}',
        },
        {
          Name: '/k8s/my-service/ARRAY_SECRET',
          Value: '["ITEM1","ITEM2"]',
        },
        {
          Name: '/k8s/my-service/JSON_URLS',
          Value: `{ ${[
            '"string": "https://example.com/slug"',
            '"object": {"key": "https://example.com/slug"}',
            '"array": ["https://example.com/slug", "https://example.com/slug"]',
          ].join(', ')}}`,
        },
      ],
    })),
  })),
}))

describe('infra CLI', () => {
  beforeEach(() => {})
  describe('render-secrets command', () => {
    beforeEach(async () => {
      const myServiceSetup = (): ServiceBuilder<'my-service'> =>
        service('my-service').secrets({
          A: '/k8s/my-service/A',
        })

      Charts.islandis = {
        dev: [myServiceSetup()],
        staging: [],
        prod: [],
      }
    })

    it('should fetch all secrets when there are some', async () => {
      // Arrange
      const argv = { service: 'my-service' }

      // Act
      const secretsMap = await renderSecretsCommand(argv.service)

      // Assert
      expect(secretsMap).toStrictEqual({ A: 'B' })
    })

    it('should fetch 0 secrets when there are none', async () => {
      // Arrange
      const argv = { service: 'emptyService' }

      // Act
      const secretsMap = await renderSecretsCommand(argv.service)

      // Assert
      expect(secretsMap).toStrictEqual({})
    })
  })

  describe('render-local-env', () => {
    beforeEach(async () => {
      const myServiceSetup = () =>
        service('my-service')
          .env({
            REMOTE_URL_ENV: 'https://www.website.tld/some/slug-path',
          })
          .secrets({
            JSON_SECRET: '/k8s/my-service/JSON_SECRET',
            ARRAY_SECRET: '/k8s/my-service/ARRAY_SECRET',
            JSON_URLS: '/k8s/my-service/JSON_URLS',
          })
      const myService = myServiceSetup()
      Charts.islandis = {
        dev: [myService],
        staging: [],
        prod: [],
      }
    })

    // Not fully implemented, not design chosen
    it.skip('should map all URIs to localhost', async () => {
      // Arrange
      const argv = {
        services: ['my-service'],
      }

      // Act
      const result = await renderLocalServices(argv)

      // Assert
      expect(result.services['my-service'].env['REMOTE_URL_ENV']).toMatch(
        /http:\/\/localhost(:\d+)?\/some\/slug-path/,
      )
      expect(result).toMatchSnapshot()
    })

    it.skip('should map all secrets to localhost', async () => {
      // Arrange
      const argv = {
        services: ['my-service'],
        withSecrets: SecretOptions.withSecrets,
      }

      // Act
      const result = await renderLocalServices(argv)

      // Assert
      expect(
        JSON.parse(result.services['my-service'].env['JSON_SECRET']),
      ).toStrictEqual({ key: ['value'] })
      // localhost rewrite
      expect(
        JSON.parse(result.services['my-service'].env['JSON_URLS'])?.array,
      ).toStrictEqual([
        'http://localhost/some/slug',
        'http://localhost/some/slug',
      ])
      expect(result).toMatchSnapshot()
    })
  })
})
