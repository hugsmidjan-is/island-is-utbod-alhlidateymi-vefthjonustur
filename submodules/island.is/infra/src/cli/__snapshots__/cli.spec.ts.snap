// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`infra CLI render-local-env should map all URIs to localhost 1`] = `
Object {
  "mocks": "docker run --name mountebank -it --rm -p 2525:2525 -v /var/home/person/islandis/island.is/infra/mountebank-imposter-config.json:/app/default.json:z docker.io/bbyars/mountebank:2.8.1 start --configfile=/app/default.json",
  "services": Object {
    "my-service": Object {
      "commands": Array [
        "cd \\".\\"",
        ". ./.env.my-service",
        "echo \\"Starting my-service in $PWD\\"",
        "yarn nx serve my-service",
      ],
      "env": Object {
        "ARRAY_SECRET": "[\\"ITEM1\\",\\"ITEM2\\"]",
        "JSON_SECRET": "{\\"key\\": [\\"value\\"]}",
        "JSON_URLS": "{ \\"string\\": \\"http://localhost/slug\\", \\"object\\": {\\"key\\": \\"http://localhost/slug\\"}, \\"array\\": [\\"http://localhost/slug\\", \\"http://localhost/slug\\"]}",
        "PROD_MODE": "true",
        "REMOTE_URL_ENV": "http://localhost/some/slug-path",
        "SERVERSIDE_FEATURES_ON": "",
      },
    },
  },
}
`;
